[BlackJack - Step1] Submit Bojana's Mission

bojana-petroska opened the pull request:
### BlackJack Step 1

This PR implements the BlackJack game, structured into the following components: `Model`, `Controller`, `InputView`, and `OutputView`.

---

### Features Implemented

### Model

- **Participant** (abstract class): Contains shared logic between `Player` and `Dealer`.
- **Player**: Is inherited from `Participant`.
- **Dealer**: Is inherited from `Participant` and adds:
  - Deck management.
  - Comparing game results with each player.
- **Card**: Represents a single card, using `Suit` and `Rank` enums.
- **Deck**: Holds and shuffles 52 standard cards.
- **Hand**: Belongs to each `Participant` and manages their cards.
- **GameResult**: Tracks each participant’s outcome.

---

### Controller

- **BlackJackGame**: Coordinates the game flow and connects model and view layers.

---

### View

- **InputView**: Handles and validates user input.
- **OutputView**: Displays game output to the user.

---

### Feedback Requested

We’d appreciate your thoughts on the following:

1. **Abstraction Design**: We used an abstract class (`Participant`) for shared logic between `Player` and `Dealer`.  
   - Do you think this is a good use of an abstract class?  
   - Would using an interface have been a better design choice in this context?

2. **Single Responsibility Principle**:  
   - Does each class follows well the Single Responsibility Principle?  
   - Are there any responsibilities that feel overloaded or misplaced?

3. **GameResult Management**:  
   - Does storing game results inside each `Participant` make sense from a design perspective, or would a centralized game statistics class would be better?

---

Thanks so much for your time and feedback!  
— *Bojana*

SebasNadu committed:
docs(readme): create the initial feature list

SebasNadu committed:
feat(blackjack): add main application entry point

SebasNadu committed:
test(model): add PlayerTest, to check the initializer and to check if the player has blackJack or busts

SebasNadu committed:
feat(model): add Participant abstract class and Player classes that inherit from Participant

SebasNadu committed:
feat(model): add Card Class with Suit and Rank enums and Card test

implement the Suit Enum with the card's symbol and the Rank Enum the number's symbol and their value

SebasNadu committed:
feat(model): add Deck Class that contains a list of cards and create a DeckTest

it constructs a deck by creating a list of 52 shuffled cards

SebasNadu committed:
feat(model): add Hand class to manage dealt cards and HandTest for validation

implement Hand class to track and evaluate card scores, including methods for blackjack and bust detection. add HandTest to validate card scoring logic and functionality.

SebasNadu committed:
docs(readme): update feature list to reflect recent model implementation progress

SebasNadu committed:
refactor(model): delegate card and score management to Hand class

Refactor `Participant` to use the `Hand` class for card and score management. Adjust related tests to validate `Hand` functionality. Simplified logic in `Participant` for cleaner design.

SebasNadu committed:
feat(model): add a Dealer Class inherited from the abstract Participant Class and add dealCard and shouldStand methods

SebasNadu committed:
feat(input): add getPlayersNames and getAnswer

SebasNadu committed:
feat(output): add printParticipantsHands and getPlayersNames

SebasNadu committed:
feat(controller): add start of the game, createPlayers and dealFirstHands

SebasNadu committed:
feat(model): add override toString logic in Dealer

SebasNadu committed:
refactor(model): change property from private to protected

SebasNadu committed:
docs(readme): update progress on Player and Dealer class features

SebasNadu committed:
refactor(model): rename methods for clarity in Hand and Dealer classes

Renamed `isBlackjack` to `hasBlackJack` in `Hand` and `shouldStand` to `shouldNotStand` in `Dealer` for improved readability. Updated related test cases accordingly.

SebasNadu committed:
refactor(input): add player name to getAnswer prompt for clarity

SebasNadu committed:
feat(output): add methods to print dealer's actions and final hands

Add `printDealersMessage` to display dealer's drawn card message and `printFinalHands` to show final hands with scores of players and dealer.

SebasNadu committed:
feat(controller): implement game logic to determine outcomes

Add methods to handle card dealing for players and dealer. Implement logic to calculate game outcomes, determining wins, losses, and ties based on scores.

SebasNadu committed:
feat(controller): implement game logic to determine outcomes

Add methods to handle card dealing for players and dealer. Implement logic to calculate game outcomes, determining wins, losses, and ties based on scores.

SebasNadu committed:
refactor(controller): extract result calculation logic into separate methods

Moved game outcome determination logic into `calculateResults` and `setResultFrom` methods for improved readability and maintainability. Encapsulated dealer's card visibility with `showAllCards` method.

SebasNadu committed:
docs(readme): update implementation progress for Controller and View components

SebasNadu committed:
feat(model): add getResults method to Participant and overrides in Player, Dealer

Introduce abstract `getResults` method in `Participant`, with custom implementations in `Player` and `Dealer` to format game results. Add `setResultFor` in Dealer to determine outcomes against players.

SebasNadu committed:
refactor(controller): streamline result calculation using Dealer's setResultFor

Simplify `calculateResults` by delegating player outcome determination to Dealer's `setResultFor` method. Remove redundant `setResultFrom` logic and enhance readability.

SebasNadu committed:
feat(output): add printResults method to display final game results

Introduce `printResults` in `OutputView` to showcase the final results for both the dealer and players, enhancing game result presentation.

SebasNadu committed:
feat(view, controller): enhance input validation and dealer message handling

Add non-blank validation for player names in `InputView`. Introduce retry mechanism in `BlackJackGame` for resilient input prompts. Refactor dealer message handling in `OutputView` by splitting into `printDealersDrawMessage` and `printDealersStandMessage` for improved clarity.

SebasNadu committed:
refactor(deck, docs): improve deck exception handling and update README progress

Update `Deck` to provide detailed exception messages when no cards are available. Mark completed tasks in the README, including user input validation with retry logic and game step formatting.

bojana-petroska committed:
init

SebasNadu committed:
docs(readme): update todo section with new tasks for exceptions, prompts, and card colors

bojana-petroska committed:
Merge remote-tracking branch 'sebas/step1' into step1

bojana-petroska committed:
style: format all the files following kotlin convention rules

da-nyee commented on README.md:
```diff
@@ -1 +1,58 @@
+
+## Features
+
+### Model
```
Well documented 👍

da-nyee commented on README.md:
```diff
@@ -1 +1,58 @@
+  - [x] Should print the Final Result.
+  - [x] Add empty lines between each step on the game.
+
+### Todo
+  - [ ] Add customize exception messages
+  - [ ] Add constants to Input and Output Prompts
+  - [ ] Print cards with color (black and red)
+  - [ ] Update the hasBlackJack method to return true only in the first hand
```
Are you planning to handle these later?

bojana-petroska replied:
yes, we wanted to have a to do list and hope we will manage it given the time constraints!

da-nyee commented on src/main/kotlin/blackjack/controller/BlackJackGame.kt:
```diff
@@ -0,0 +1,84 @@
+
+class BlackJackGame {
+    val dealer = Dealer()
+    var players = listOf<Player>()
```
How about wrapping this collection?

bojana-petroska replied:
Thank you for the suggestion, it really makes sense. I will implement a new Collection for Players.

da-nyee commented on src/main/kotlin/blackjack/controller/BlackJackGame.kt:
```diff
@@ -0,0 +1,84 @@
+        dealFirstCards()
+        dealingPlayersCards()
+        dealingDealersCards()
+        calculateResults()
```
Nice method extraction — it’s clear what each method is doing 👍

da-nyee replied:
However, consider making them `private`, since they’re only used within this class.

da-nyee commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -0,0 +1,30 @@
+package blackjack.model
+
+abstract class Participant(val name: String) {
```
Yes! I think using an `abstract class` is better than an `interface` for now.
If you need a higher level of abstraction later, you can consider converting it to an `interface`.
I’d recommend approaching abstraction step-by-step — from a concrete `class`, to an `abstract class`, and then to an `interface`.

bojana-petroska replied:
Thanks! This is a very useful hint, I have spent the whole week wondering about in which case an abstract class is beneficial and in which an interface : )

da-nyee commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,37 @@
+package blackjack.model
+
+class Hand {
```
Good name choice 👍

da-nyee commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -0,0 +1,11 @@
+        return when {
+            gameResults.wins > 0 -> "$name: Win"
+            gameResults.loses > 0 -> "$name: Lose"
+            else -> "$name: Tie"
```
I guess the model now depends on the view — how about moving this logic to the `OutputView` instead?
`Model` should stay independent from `Controller` and `View`.

bojana-petroska replied:
Yes, makes sense. I did it for both, the Player and the Dealer.

da-nyee commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,71 @@
+        val playerScore = player.getScore()
+        val dealerScore = this.getScore()
+        when {
+            player.isBusts() -> {
+                player.setLose()
+                this.setWin()
+            }
```
In Kotlin, there’s a useful function called `apply` that you can use when setting values to properties.
It helps make your code more concise and improves readability by grouping related property assignments together.

bojana-petroska replied:
do you mean like this:
```
player.isBusts() -> {
     player.apply { setLose() }
     this.apply { setWin() }
}
```

bojana-petroska replied:
I used it in the tests for some of the chained card additions

da-nyee replied:
@bojana-petroska 

```kotlin
player.isBusts() -> apply {
    player.setLose()
    this.setWin()
}
```

I meant like this !

da-nyee commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,71 @@
+    override fun getResults(): String {
+        val winText = if (gameResults.wins > 0) "${gameResults.wins} Win" else ""
+        val loseText = if (gameResults.loses > 0) "${gameResults.loses} Lose" else ""
+        val tieText = if (gameResults.ties > 0) "${gameResults.ties} Tie" else ""
```
I'd recommend moving this logic to the view as well, as I mentioned.

da-nyee commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,29 @@
+object InputView {
+    fun getPlayersNames(): List<String> {
+        println("Enter the names of the players (coma-separated):")
+        val input = readLine() ?: ""
```
```suggestion
        val input = readlnOrNull() ?: ""
```

Consider using this method to prevent a `NullPointerException`, which is one of the most problematic runtime exceptions.

bojana-petroska replied:
good suggestion, I actually used it in Lotto : )))

da-nyee commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,29 @@
+
+    fun getAnswer(playerName: String): String {
+        println("Would $playerName like to draw another card? (y for yes, n for no)")
+        val input = readLine()?.trim()?.lowercase() ?: ""
```
Here too.

da-nyee commented on src/test/kotlin/blackjack/model/PlayerTest.kt:
```diff
@@ -0,0 +1,25 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+class PlayerTest {
```
I'd suggest adding more tests, such as `DealerTest`, `HandTest`, and so on.

da-nyee reviewed (changes requested):
Hi Bojana! 🙌
You did a really good job — nice work 👍 
It seems like you have a solid understanding of OOP and how to apply it in your code.
I left some comments — feel free to check them out and let me know if you have any questions 👀

bojana-petroska committed:
refactor(controller, model): add Players class to encapsulate player related logic

- Moved logic related to managing multiple players from BlackJackGame into a new Players class.
- This improves separation of concerns and prepares the codebase for easier expansion and testing.

bojana-petroska committed:
refactor(model, view): move result formatting from model to OutputView

- Removed getResults() from Player and Dealer model classes to keep models free from view logic
- Added printPlayersResults() and printDealerResults() to OutputView to handle result display
- Updated BlackJackGame to call OutputView for printing results

bojana-petroska committed:
refactor(model, view): move card display logic from model to OutputView for better separation of concerns

- Removed toString override from Participant.
- Updated OutputView to handle dealer and player hand formatting explicitly.

bojana-petroska committed:
refactor(view): add readlnOrNull instead of readLine to prevent a NullPointerException

bojana-petroska committed:
test(model): add unit tests for Dealer class

da-nyee commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,90 @@
+    fun printPlayersResults(players: List<Player>) {
+        players.forEach { player ->
+            val results = when {
+                player.gameResults.wins > 0 -> "${player.name} Win"
+                player.gameResults.loses > 0 -> "${player.name}: Lose"
```
You moved them 👍

da-nyee commented on src/test/kotlin/blackjack/model/HandTest.kt:
```diff
@@ -0,0 +1,50 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class HandTest {
```
👍

da-nyee reviewed (approved):
Hi Bojana! 🙌
Thank you for applying the feedback 🙇‍♀️

I went through all the changes, and I think it's good to merge this now, since we have one more step to go.
I also left a comment — feel free to take a look. Let me know if you have any other questions.

Cheers, and good luck with the next step 💪
